trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    // Invalid chunk size
+    @Test
+    public void testCorruptChunkedInputStreamInvalidSize()
+            throws IOException
+    {
+        String s = "whatever;chunk-signature=0\r\n01234\r\n5;chunk-signature=0\r\n56789\r\n0;chunk-signature=0\r\n\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        assertThrows(IOException.class, in::read);
+        in.close();
+    }
+
+    // Negative chunk size
+    @Test
+    public void testCorruptChunkedInputStreamNegativeSize()
+            throws IOException
+    {
+        String s = "-5;chunk-signature=0\r\n01234\r\n5;chunk-signature=0\r\n56789\r\n0;chunk-signature=0\r\n\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        assertThrows(IOException.class, in::read);
+        in.close();
+    }
+
+    // Truncated chunk
+    @Test
+    public void testCorruptChunkedInputStreamTruncatedChunk()
+            throws IOException
+    {
+        String s = "3;chunk-signature=0\r\n12";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        byte[] buffer = new byte[300];
+        assertEquals(2, in.read(buffer));
+        assertThrows(IOException.class, () -> in.read(buffer));
+        in.close();
+    }
+
+    @Test
+    public void testCorruptChunkedInputStreamClose()
+            throws IOException
+    {
+        String s = "whatever;chunk-signature=0\r\n01234\r\n5;chunk-signature=0\r\n56789\r\n0;chunk-signature=0\r\n\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(UTF_8));
+        try (InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession())) {
+            assertThrows(IOException.class, in::read);
+        }
+    }
+
+    @Test
+    public void testEmptyChunkedInputStream()
+            throws IOException
+    {
+        String s = "0;chunk-signature=0\r\n\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        byte[] buffer = new byte[300];
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        int len;
+        while ((len = in.read(buffer)) > 0) {
+            out.write(buffer, 0, len);
+        }
+        assertEquals(0, out.size());
+        in.close();
+    }
+
+    // Test for when buffer is larger than chunk size
+    @Test
+    public void testHugeChunk()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream("499602D2;chunk-signature=0\r\n01234567".getBytes(UTF_8));
```

Comment:
This is the one I'm conflicted about - we check that:
- All streams must have a valid final chunk
- All chunk signatures are valid prior to streaming the last byte in the chunk

We have a test to verify that if a chunk contains more data than its header indicates, we'll throw.

However, if the header indicates the chunks are larger than they actually are we are potentially allowing unsigned content to go through.

I'm wondering whether we should have `AwsChunkedInputStream` check the `X-Amz-Decoded-Content-Length` and force a validation prior to reading the last byte. Otherwise we could be in a situation where someone can send in a request with `aws-chunked` encoding and a much larger chunk size than `X-Amz-Decoded-Content-Length` - we will never validate the signature of the chunk

Reply:
Your concern is processing more bytes than `X-Amz-Decoded-Content-Length` says we should?

Reply:
My concern is having a single chunk with a header stating a much larger chunk size than X-Amz-Decoded-Content-Length and some invalid signature

this chunk can contain X-Amz-Decoded-Content-Length bytes (which as stated would be much smaller than the chunk size stated in the chunk header)

we would stream all bytes to S3 until we get an EOF, but by this point S3 has processed the payload already without any signature verification

Reply:
it would be the same with 2 chunks. Chunk one has a reasonable size and the second a huge size. I think if we have `AwsChunkedInputStream` take an optional max bytes argument we can prevent this.

Reply:
However, this means we need to track `X-Amz-Decoded-Content-Length` (probably in `RequestContent`). I suggest that we do this as a follow-up PR. OK?

Reply:
https://github.com/trinodb/aws-proxy/issues/106